### DISCUSSION TITLE
[Documentation] How to create a service file to run at boot, in the init distrobox.

### DISCUSSION THREAD

**Discussion by Aman9das** [89luca89/distrobox#1363] [General]
**Is your feature request related to a problem? Please describe.**
I want to install [ollama.ai](https://github.com/jmorganca/ollama/blob/main/docs/linux.md) in a systemd container, but for some reason the service doesn't start when entering the distrobox. The container was created with the `-I` flag.

**Describe the solution you'd like**
An addition to the documentation near the init containers documentation, about how to make sure a service runs at container bootup.

**Describe alternatives you've considered**
I've tried to specify the container is `After=default.service` but it doesn't run on boot :(

**Additional context**
https://www.freedesktop.org/software/systemd/man/latest/systemd.special.html
https://www.howtogeek.com/687970/how-to-run-a-linux-program-at-startup-with-systemd/

**Comment by pavinjosdev** [comment]
I have a systemd user service file for starting an init distrobox container. Perhaps it's applicable for your use as well.

Create file `/etc/systemd/user/distrobox.service` with the following contents:

```
[Unit]
Description=Start distrobox containers

[Service]
Type=oneshot
# Without some delay, distrobox was taking control of audio
ExecStartPre=/usr/bin/sleep 30
ExecStart=/usr/bin/podman start deb
ExecStop=/usr/bin/podman stop deb
RemainAfterExit=yes

[Install]
WantedBy=default.target
```

Replace `deb` above with your own container name

Verify config:

```
systemd-analyze verify /etc/systemd/user/distrobox.service
```

Enable and start it:
```
systemctl --user enable --now distrobox.service
```

**Comment by kilesduli** [comment]
> I have a systemd user service file for starting an init distrobox container. Perhaps it's applicable for your use as well.
> 
> Create file `/etc/systemd/user/distrobox.service` with the following contents:
> 
> ```
> [Unit]
> Description=Start distrobox containers
> 
> [Service]
> Type=oneshot
> # Without some delay, distrobox was taking control of audio
> ExecStartPre=/usr/bin/sleep 30
> ExecStart=/usr/bin/podman start deb
> ExecStop=/usr/bin/podman stop deb
> RemainAfterExit=yes
> 
> [Install]
> WantedBy=default.target
> ```
> 
> Replace `deb` above with your own container name
> 
> Verify config:
> 
> ```
> systemd-analyze verify /etc/systemd/user/distrobox.service
> ```
> 
> Enable and start it:
> 
> ```
> systemctl --user enable --now distrobox.service
> ```

I thought there's a better solution instead of sleep. 
Simply set the `After` and `Wants` directives to `org.gnome.Shell@x11.service` or another appropriate service that triggers after user login. 
This service should come after <xxxx>-session-manager.service. 
For SDDM, consider searching for services with `@x11` or `@wayland` suffix.

**Comment by pavinjosdev** [comment]
> I thought there's a better solution instead of sleep.
> Simply set the `After` and `Wants` directives to `org.gnome.Shell@x11.service` or another appropriate service that triggers after user login.

That's a great idea! 🤗
I think `After=default.target` would be more portable, IIRC that worked for me.
But I ended up keeping the sleep part because my distrobox volumes/mounts depend on some slow starting external USB drives (LUKS encrypted and pooled together as btrfs RAID volume) that are part of the system service manager.

I have no idea how to reconcile the distrobox service file in the user service manager to depend on the startup of services from the system manager. 🤔

EDIT 1: I ended up moving the distrobox startup to the system service manager:

```
# /etc/systemd/system/container-autostart.service
[Unit]
Description=Start containers on system startup
After=usbraid.mount

[Service]
Type=oneshot
User=pavin
Group=pavin
ExecStart=/usr/bin/podman start deb jellyfin
ExecStop=/usr/bin/podman stop deb jellyfin
RemainAfterExit=yes

[Install]
WantedBy=graphical.target
```

**Comment by archerallstars** [comment]
I use a `--user` systemd service to start a Distrobox container at boot. I find the user's systemd service to be simple and safe to use, for example in my `~/.config/systemd/user/cfw-dbx.service`:

```
[Unit]
Description=Start cfw-dbx container for Cloudflare WARP connection.

[Service]
Type=oneshot
ExecStart=-bash -c "distrobox enter cfw-dbx"
ExecStop=-bash -c "distrobox stop cfw-dbx"
Restart=on-failure
RestartSec=1s
RemainAfterExit=yes

[Install]
WantedBy=default.target
```

I use `Restart=on-failure` since it's hard to pinpoint which service, socket, component, etc. will be ready for Distrobox. I tried with `After` and `Requires` in a system service file using `default.target`, `podman.socket`, a service that checks for `crun` to be available, but none of them worked for me.

I put `RestartSec=1s` so it won't spam my login too much.

**Comment by 89luca89** [comment]
Converting to discussion for future reference :+1:

**Comment by ericcurtin** [comment]
Just saying if you use this tool:

https://github.com/ericcurtin/podman-ollama

it's:

```
podman-ollama -p 11434:11434 generate systemd --new --files --name podman-ollama; cat /home/ecurtin/git/podman-ollama/container-podman-ollama.service
/home/ecurtin/git/podman-ollama/container-podman-ollama.service
# container-podman-ollama.service
# autogenerated by Podman 4.3.1
# Wed May  1 20:05:21 IST 2024

[Unit]
Description=Podman container-podman-ollama.service
Documentation=man:podman-generate-systemd(1)
Wants=network-online.target
After=network-online.target
RequiresMountsFor=%t/containers

[Service]
Environment=PODMAN_SYSTEMD_UNIT=%n
Restart=on-failure
TimeoutStopSec=70
ExecStartPre=/bin/rm \
        -f %t/%n.ctr-id
ExecStart=/usr/bin/podman run \
        --cidfile=%t/%n.ctr-id \
        --cgroups=no-conmon \
        --rm \
        --sdnotify=conmon \
        --replace \
        -p 11434:11434 \
        --privileged \
        --security-opt=label=disable \
        --name podman-ollama \
        -d \
        --gpus=all \
        --device /dev/dri \
        -v/home/ecurtin:/home/ecurtin \
        -vollama:/root/.ollama docker.io/ollama/ollama:latest
ExecStop=/usr/bin/podman stop \
        --ignore -t 10 \
        --cidfile=%t/%n.ctr-id
ExecStopPost=/usr/bin/podman rm \
        -f \
        --ignore -t 10 \
        --cidfile=%t/%n.ctr-id
Type=notify
NotifyAccess=all

[Install]
WantedBy=default.target
```
